File: Stacks/valid_parentheses.py
"""
https://leetcode.com/problems/valid-parentheses/description/
"""
class Solution: #0(n) time complexity #O(n) space complexity
    def isValid(self, s: str) -> bool:
        stack = []
        pairs = {
                "{": "}", 
                "(":")",
                "[":"]"
                }

        for i in range(len(s)):
            if s[i] in pairs:
                stack.append(pairs[s[i]])
            else:
                if not stack or s[i] != stack.pop(): # "if stack and s[i] != stack.pop()" wouldn't work for the edge case of a single bracket (e.g "]")
                    return False

        return len(stack) == 0

File: Stacks/valid_parentheses.js
/*
https://leetcode.com/problems/valid-parentheses/description/
*/

class Solution { //Time complexity 0(n), Space complexity 0(n)
    /**
     * @param {string} s
     * @return {boolean}
     */
    isValid(s) {
        const hashmap = {")": "(", "]":"[", "}":"{"}
        let stack = []

        for (const c of s){
            if (!(c in hashmap)){
                stack.push(c)
            }else{
                if(!stack.length || hashmap[c] != stack[stack.length - 1]){
                    return false
                }else{
                    stack.pop()
                }
            }
        }

        return !stack.length
    }
}
